### PR TITLE
Modded Leaderboard Patch

### PR DESCRIPTION
**Pull request by Xilophor**
Creates a separate modded leaderboard and modded challenge save file (see #2)

### CODE REVIEW

**Comment by legoandmars** [comment]
looks good, seems to work from my testing 

I might add a (Modded) or some other indication this is separate from the Vanilla challenge at the top text on the leaderboard menu
image

**Comment by legoandmars** [comment]
In `LobbyCompatibilityPlugin.GetAsyncInfo`, I would replace `(BindingsFlags)60` with `BindingFlags.Instance | BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic`, just so it's slightly clearer what it's doing

You might be able to just replace it with `BindingFlags.Instance | BindingFlags.NonPublic`

**Comment by MaxWasUnavailable** [comment]
@Xilophor Needs merge conflict fixing

File: LobbyCompatibility/LobbyCompatibilityPlugin.cs [hunk 1]
```diff
@@ -26,6 +30,25 @@ private void Awake()
...
+    
+    /// <summary>
+    /// Retrieves the MethodInfo of the compiler-generated async method.
+    /// Async method content is not actually in the <c>async Method()</c>, but instead is in a separate <c>struct</c> under the method "MoveNext";
+    /// this function retrieves that method info.
+    /// </summary>
+    /// <param name="type">(<see cref="Type"/>) The type of the class housing the method.</param>
+    /// <param name="method">(<see cref="string"/>) The name of the method being patched.</param>
+    /// <returns>(<see cref="MethodInfo"/>) The info of the async "MoveNext" method.</returns>
+    public static MethodInfo? GetAsyncInfo(Type type, string method)
```
Comment (MaxWasUnavailable) [inline LobbyCompatibility/LobbyCompatibilityPlugin.cs:43]: Can this be moved to a helper? E.g. "HarmonyHelper"?

File: LobbyCompatibility/Patches/ChallengeSaveFilePatch.cs [hunk 2]
```diff
@@ -1,16 +1,16 @@
...
 
 [HarmonyPatch]
-[HarmonyPriority(Priority.First)]
+[HarmonyPriority(Priority.Last)]
 [HarmonyWrapSafe]
 internal class ChallengeSaveFilePatch
 {
     [HarmonyPrefix]
-    [HarmonyPatch(typeof(ES3Settings), MethodType.Constructor, typeof(string), typeof(ES3Settings))]
+    [HarmonyPatch(typeof(ES3Settings), MethodType.Constructor, new Type[] { typeof(string), typeof(ES3Settings) })]
```
Comment (MaxWasUnavailable) [inline LobbyCompatibility/Patches/ChallengeSaveFilePatch.cs:11]: Explicit "new Type[]" isn't required

File: LobbyCompatibility/Patches/ChallengeSaveFilePatch.cs [hunk 3]
```diff
@@ -1,16 +1,16 @@
...
     private static void FileSettings(ref string path, ES3Settings settings)
     {
-        path = path == "LCChallengeFile" ? "LCModdedChallengeFile" : path;
+        path = (path == "LCChallengeFile") ? "LCModdedChallengeFile" : path;
```
Comment (MaxWasUnavailable) [inline LobbyCompatibility/Patches/ChallengeSaveFilePatch.cs:14]: () isn't required here

File: LobbyCompatibility/Patches/LeaderboardNamePatch.cs [hunk 4]
```diff
@@ -1,26 +1,17 @@
...
 internal class LeaderboardNamePatch
 {
-    [HarmonyTargetMethod]
-    private static MethodBase TargetMethod()
-    {
-        return typeof(SteamUserStats).Assembly.GetType("Steamworks.ISteamUserStats")!.GetMethod(
-            "FindOrCreateLeaderboard", (BindingFlags)60)!;
-    }
-
+    [HarmonyPatch(typeof(ISteamUserStats), nameof(ISteamUserStats.FindOrCreateLeaderboard))]
```
Comment (MaxWasUnavailable) [inline LobbyCompatibility/Patches/LeaderboardNamePatch.cs:11]: For consistency with other patches, I'd probably rename the class to the harmony target class, and add [HarmonyPatch(...)] to the class, with the method HarmonyPatch to the method.

**Review by MaxWasUnavailable — CHANGES_REQUESTED** [review]

**Review by MaxWasUnavailable — APPROVED** [review]

**Pull request merged**